### ISSUE TITLE
Question: One comment on making the past works work with a modern compiler

### ISSUE THREAD

**Issue by mame** [ioccc-src/temp-test-ioccc#879]
I now realize that you have been trying to make past works buildable by modern compilers.

I do never mean to deny your efforts, but I personally like to keep the original code untouched.

I don't think that it is desirable to destroy the essence of each era's obfuscation in order to make it work with modern compilers. As IOCCC's past works have proven, the C language is changing. Even if it works with the compiler of this very moment, it is likely that it will not work later. If we make more modifications at such a time, we may even lose the essence of the work.

I wrote the patches on my spoiler site so that people at this very moment can check the behavior of the past works, and never thought that the original code should be altered.

I am not asking you to revert changes you have already made so far. However, I do have one request: please keep the original version with a consistent name. I think it should be AUTHOR_NAME.c, but AUTHOR_NAME.orig.c is also fine for me.

By the way, I personally would really like you to open the next contest (not a MOCK, but a real one) rather than investing your effort in the past works. My `ioccc2021/` folder just keeps adding weird code! ;-)

**Comment by lcn2** [comment]
Thank you @mame for your comments.  We do appreciate them.

Your idea of providing a `foo.orig.c` in each entry is a good idea.  We can, in fact, do that now even for entries that might not have been modified as a precaution.

In some cases, entries come with a fair number of additional files, scripts and alternate source.   The `foo.orig.c` would only be a copy of the original C code as some entries such as `2020/ferguson1/` come with a bunch of files and duplicating each and every file might be a bit much.  Moreover, in many cases, the Makefile had to be modified from the original: whose in early years there was one or two Makefiles for the entire year: early years were just a set of files in a single year directory and now each winner has their own directory.

So while we can providing a `foo.orig.c` file for all entries, we cannot hope to to compile such code.  However ...

The web site will has an `archive/` directory containing archived compressed tarballs going back to 2020 Dec 28.  See the  [archive/README.md](https://github.com/ioccc-src/temp-test-ioccc/blob/master/archive/README.md) for details.  This set will allow someone to rebuild a given a year if they so wish.  In addition we have added a draft of [www-history.md](https://github.com/ioccc-src/temp-test-ioccc/blob/master/www-history.md) (currently under development) that discusses the history of the [official IOCCC web site](https://www.ioccc.org) and makes references to the [IOCCC web site has evolved](https://web.archive.org/web/20230000000000*/www.ioccc.org) where several thousand snapshots of the site may be found for those truly interested in recreating history.  We have made an effort to be sure that the [Internet Wayback Machine Wayback Machine](https://web.archive.org) kept a good track of changes to the [official IOCCC web site](https://www.ioccc.org).

Comments, concern and suggestions are welcome.

**Comment by lcn2** [comment]
While we have your attempting @mame, if you believe that you have a better modification to an entry that is currently found in the [temp-test-ioccc repo](https://github.com/ioccc-src/temp-test-ioccc/), please feel free to recommend it.

As you noted, our model is to have `make all` try to do a reasonable thing with all entries.  And as mentioned sometimes there is a `make alt` that provides an alternative, however in some cases the alternative is **NOT** the original, in some cases the original entry came with an alternative code, etc.

Even so, feel free to make suggestions for better ways to make old code work under more modern systems by default.

**Comment by lcn2** [comment]
> I now realize that you have been trying to make past works buildable by modern compilers.
> 
> I do never mean to deny your efforts, but I personally like to keep the original code untouched.
> 
> I don't think that it is desirable to destroy the essence of each era's obfuscation in order to make it work with modern compilers. As IOCCC's past works have proven, the C language is changing. Even if it works with the compiler of this very moment, it is likely that it will not work later. If we make more modifications at such a time, we may even lose the essence of the work.
> 
> I wrote the patches on my spoiler site so that people at this very moment can check the behavior of the past works, and never thought that the original code should be altered.
> 
> I am not asking you to revert changes you have already made so far. However, I do have one request: please keep the original version with a consistent name. I think it should be AUTHOR_NAME.c, but AUTHOR_NAME.orig.c is also fine for me.
> 
> By the way, I personally would really like you to open the next contest (not a MOCK, but a real one) rather than investing your effort in the past works. My `ioccc2021/` folder just keeps adding weird code! ;-)

We understand the desire to get the next IOCCC started.  As you well know, the IOCCC was perpetually late in getting the winning entries released.  And "perpetually late" is just a polite way of saying that past IOCCC's were frustratingly and unacceptably delayed for poor reasons.  We would rather not go into details, but let's just say that someone in the past had significant problems finishing even simple tasks.  Moreover, they put into the system, a rather complex and fragile patchwork of procedures that made it extremely difficult for the IOCCC judges to complete things without the manual invention of this other person.  And the problem was becoming worse and worse to the point where the tools were unmaintainable and the ability to publish winners without that other person finishing tasks was impossible.

We really would rather not go into details about the above person: we would rather keep their name "out of the press" and not impugn them in any way.  So even if you can guess who it might be, we ask that you not say so.  They are longer involved with the IOCCC, and we wish them well.

The 2020 winners were "99% ready the day after they were announced" and all that was needed was for that person to simply type in one command to publish the 2020 winners.  While it would not have been a perfect release of the 2020 code, it was good enough for others to enjoy and good enough to honor those who work hard to submit entries.  After many months of no action on the part of that other person, the IOCCC judges had to act.  Fortunately, we owned the _ioccc.org_ domain.  Moreover the nice folks at GitHub were kind enough to host the [official IOCCC web site](https://www.ioccc.org).  They have the distributed servers to handle the **HUGE** load that comes when a new IOCCC winner set is released.  So, to publish the 2020 winners, we had to build by hand the [IOCCC winner repo](https://github.com/ioccc-src/winner) and then flip the DNS record to use it, along with sending the MX to a new spot, etc. etc. etc.

We do not want to go down that slow road again.  Instead, we plan to use git via GitHub to publish winners.  Announcing the winners **will be as simple as a git push of the new year's directory**.  Winners, and other will be able to suggest ways to improve, fix, correct, and update writeup via GitHub pull requests.  The web pages will be automatically built by tools that will be part of the GitHub repo.

Authors will be able to modify their information about themselves by modifying a simple JSON file.  For an example draft, see [author/Yusuke_Endoh.json](https://github.com/ioccc-src/temp-test-ioccc/blob/master/author/Yusuke_Endoh.json).  Those JSON files will be used by tools to automatically build the critical web pages for the site.

Details about a winning entry will be found in a "_.winner.json_" file in each winning entry directory.  The details are still being worked out.  An [.winner.json example](https://github.com/ioccc-src/temp-test-ioccc/blob/master/tmp/example.dot_winner.json) is available.

Authors and others will be able to submit GitHub pull requests to modify such files.  And of course, we won't blindly take any pull request, only the good ones.  :-)

Markdown files will be converted into HTML, so the existing `README.md` will be used to generate `index.html` files for each winner.

When we receive a GitHub pull request, we will of course evaluate it for quality: making modifications if and as needed on a local git clone of the repo.  We will them run the master "check and build everything" script.  Then we will use a combination of `git diff` and a local web browser to inspect the result.  If all is well, we will `git push` to the repo the change.

So not only will the announcement of the winners **BE** the posting of the winning code, the process of receiving updates and improvements will be easy and rapid.

Please understand that there is *A GREAT DEAL OF EFFORT** going on beyond just the [temp-test-ioccc repo](https://github.com/ioccc-src/temp-test-ioccc/) and compiling entries.  There is folk working on the "submit server" that will allow for people to register for the IOCCC and upload their entries in the form of a compressed tarball.  That submit server must withstand the typical abuse that system crackers try to do with nice things on the Internet.  But even more important, the new submit server will allow the someone entering the IOCCC **to see the status of their submission**.   So, if for some reason, their entry is malformed, violates Rule 2a, or was otherwise invalid for some reason, the IOCCC judges can modify the author of the error.  If all is well, the author will see that their entry has been received.  Once judging has started, the author will be able to see that their entry is being judged.  And if they win, it will show up on the [official IOCCC web site](https://www.ioccc.org) via the [IOCCC winner repo](https://github.com/ioccc-src/winner).

There is, for example, a tool that will package an entry, collect information about the authors, and build a compressed tarball with all the required files ready to upload to the submit server.  The [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry) will be used by people to package their entry.   Sure, someone could package their entry themselves, but in doing so they risk having their entry rejected for some reason. At least they be notified via the submit server that the submission was invalid along with a reason why it was rejected.

And there is other "behind the screens" tools used by the IOCCC judges that are being fixed / improved as well.

The need for the **IOCCC MOCK** is to allow us to do a trail run.  We won't judge **MOCK** entries.  In fact, we will ask them to be simple "Hello, world!" entries that we can process.  If something breaks in the process, we can safely fix and restart the *IOCCC MOCK** without risking real entries.  Running the *IOCCC MOCK** should not take long.  And when all is well, we will immediately open **IOCCC28**.

So, we hope you realize that there is a great deal more effort going in in parallel to get the IOCCC ready.  We **DO PLAN** to hold another IOCCC.  We realize that it has been a while since the IOCCC last opened.  Please be assured that several people behind the scenes are working **very hard** to get the IOCCC back in order.  We are confident that when the new tool chain and process is in place, we will easily be able to hold annual IOCCC and announce winners and source quickly, and to make corrections as needed.

**Comment by mame** [comment]
Thank you for your quick reply, Landon!

## Backing up original files

Conclusion first. I will send you a PR later to put `foo.orig.c` in each directory with reference to the archive as of 2020.

I missed that files other than `.c` also need to be backed up. In fact, I don't recall needing to edit any additional files when I create my spoiler site for all entries.
If we must do that, it would be better to add a new file instead of editing an existing one. Personally, I feel that if an entry did not work at all without modification of data files other than `.c`, it would violate the first rule of IOCCC "1) Your entry must be a complete program." ;-)

I understand that it would be hard to keep Makefile as is. It's a bit of a shame because IOCCC's Makefile is also very creative, but I'm personally OK with rewriting Makefile. 

## About the contest management

I really appreciate your hard work! I guessed as much :-) I'll do my best to wait for the next contest.

The following is my personal story, which you don't have to read.

I have organized Ruby obfuscation contests called [TRICK](https://github.com/tric) four times, inspired by IOCCC.
It has whopping seven judges (including myself), so I know communication costs are harder than expected.
The hardest part to organize the contest is collecting the score sheets from all of them.
However, we managed it because it is decided that the contest result is announced at a real event called RubyKaigi.
So what IOCCC need might be a deadline :-)

On the other hand, I dare to keep the technology stack low-tech.
The submission method is email. The secretariat (my wife :-) accepts and anonymizes all submissions manually.
The public site is a GitHub repository. We don't even have GitHub Pages.
I designed this process as simple to meet the deadline, but also because the IOCCC submission site and review process seemed "too much" to me :-P

Again, thank you for your organizing the wonderful contest. I really hope that IOCCC can continue in a sustainable way.

**Comment by lcn2** [comment]
> Thank you for your quick reply, Landon!
> 
> ## Backing up original files
> Conclusion first. I will send you a PR later to put `foo.orig.c` in each directory with reference to the archive as of 2020.
> 
> I missed that files other than `.c` also need to be backed up. In fact, I don't recall needing to edit any additional files when I create my spoiler site for all entries. If we must do that, it would be better to add a new file instead of editing an existing one. Personally, I feel that if an entry did not work at all without modification of data files other than `.c`, it would violate the first rule of IOCCC "1) Your entry must be a complete program." ;-)
> 
> I understand that it would be hard to keep Makefile as is. It's a bit of a shame because IOCCC's Makefile is also very creative, but I'm personally OK with rewriting Makefile.
> 
> ## About the contest management
> I really appreciate your hard work! I guessed as much :-) I'll do my best to wait for the next contest.
> 
> The following is my personal story, which you don't have to read.
> 
> I have organized Ruby obfuscation contests called [TRICK](https://github.com/tric) four times, inspired by IOCCC. It has whopping seven judges (including myself), so I know communication costs are harder than expected. The hardest part to organize the contest is collecting the score sheets from all of them. However, we managed it because it is decided that the contest result is announced at a real event called RubyKaigi. So what IOCCC need might be a deadline :-)
> 
> On the other hand, I dare to keep the technology stack low-tech. The submission method is email. The secretariat (my wife :-) accepts and anonymizes all submissions manually. The public site is a GitHub repository. We don't even have GitHub Pages. I designed this process as simple to meet the deadline, but also because the IOCCC submission site and review process seemed "too much" to me :-P
> 
> Again, thank you for your organizing the wonderful contest. I really hope that IOCCC can continue in a sustainable way.

With commit 2ba12ecc3cdafc4688020dae061f3a6736104f6b we restored '*.orig.c' to each entry.

Under each yyyy/dir there is a file of the form '*.orig.c' containing the original source code, or as close as we can get to the original source code.  Whenever there was already some archival "foo.orig.c" file, we used that file, otherwise the "foo.c" was used as the new "foo.orig.c" file.

If you find an error in those `yyyy/dir/*.orig.c' files, such as an earlier version, let us know.

**Comment by lcn2** [comment]
> I understand that it would be hard to keep Makefile as is. It's a bit of a shame because IOCCC's Makefile is also very creative, but I'm personally OK with rewriting Makefile.

We think the new Makefiles have some creativity to them too.  :-)

**Comment by lcn2** [comment]
Please check, @mame, the new "foo.orig.c" files.  If you find an error, let us know in a comment below so we can discuss.

With that, we think this comment is ready to close: but we will wait a bit to see if you find any errors in the "foo.orig.c" files.

**Comment by mame** [comment]
Very quick action, thank you so much!

The change looks good from a quick glance. My concerns have been addressed. Let's close this issue.
I think I have a slightly older tarball copy of all entries on my old machine. I'll compare them to that later and if I notice anything I'll send a separate issue or PR.

**Comment by xexyl** [comment]
On my behalf I agree. This is kind of what I meant way back then :-)

It looks good on my end too!

Thanks @mame for some of your fixes in your wonderful analysis of the previous entries! We used some to fix those entries. I fixed many more including 2001/anonymous which you might find fun to see the changes as I know that was one that you weren't sure how to fix. This happened with some others too.

**Comment by xexyl** [comment]
> 
> By the way, I personally would really like you to open the next contest (not a MOCK, but a real one) rather than investing your effort in the past works. My `ioccc2021/` folder just keeps adding weird code! ;-)

The mock is important though as so many changes were made that it would be a real shame if a problem was found and then the entries couldn't be used or won. Funny thing is my entries that I worked on are also in a 2021 directory and if I make any new ones I'll put them there too.

We all want the next contest though but mock has to come first :-)

.. maybe Landon noted this already though .. didn't read all the comments.

**Comment by xexyl** [comment]
> Pull requests that fit into the current entry model would be very welcome, @mame. Please see [bugs.md](https://github.com/ioccc-src/temp-test-ioccc/blob/master/bugs.md) for a list of most wanted fixes.

It should be noted though that I'm working on some of these at this time. I have a work in progress at this time that I really ought to commit. Perhaps because of this comment I should get on that and if nothing else because it's annoying to have the change not in.

But help is certainly welcome even if I want to have a crack at some of the entries I added to the bugs.md file :-)

Btw to both of you, if you also don't know Landon: I have a mental record of some things that have to be done but which I have not documented in any file. This particularly goes for the README.md files. I'm kind of holding off atm on those but I will get back to those in time. That does not mean that they can't be modified but it is to say that I'm aware of issues that still have to be fixed so it's probably not worth doing it.

**Comment by xexyl** [comment]
Two more comments here and then I will continue in the other discussion you opened from now on (well and jparse thread like this one of course depending on context)... up next...

**Comment by xexyl** [comment]
![hell-for-christians-bible-study-in-hell](https://github.com/user-attachments/assets/04fb383b-d0e8-4ce4-9251-bc99c440a266)

**Comment by xexyl** [comment]
...and with that, I bid this issue a fare well!! Thanks again! It's been an honour and privilege Landon!
